lib/fluent/plugin/out_azure-storage-append-blob.rb

```diff
@@ -17,68 +18,112 @@ class AzureStorageAppendBlobOut < Fluent::Plugin::Output
+
       config_section :format do
         config_set_default :@type, DEFAULT_FORMAT_TYPE
       end
-  
+
       config_section :buffer do
-        config_set_default :chunk_keys, ['time']
+        config_set_default :chunk_keys, ["time"]
         config_set_default :timekey, (60 * 60 * 24)
       end
-  
+
       attr_reader :bs
-  
+
       def configure(conf)
         super
-  
+
         @formatter = formatter_create
-  
+
         if @localtime
-          @path_slicer = Proc.new {|path|
+          @path_slicer = Proc.new { |path|
             Time.now.strftime(path)
           }
         else
-          @path_slicer = Proc.new {|path|
+          @path_slicer = Proc.new { |path|
             Time.now.utc.strftime(path)
           }
         end
-  
+
+        if @azure_storage_account.nil?
+          raise ConfigError, "storage account needs to be specified"
+        end
+
         if @azure_container.nil?
-          raise ConfigError, 'azure_container is needed'
+          raise ConfigError, "container needs to be specified"
+        end
+
+        if !@azure_use_managed_identity && @azure_storage_access_key.nil?
+          raise ConfigError, "either managed identity is enabled or storage account key needs to be present"
         end
       end
-  
+
       def multi_workers_ready?
         true
       end
-  
+
+      def get_access_token()
+        access_key_request = Faraday.new("http://169.254.169.254/metadata/identity/oauth2/token?" +
+                                         "api-version=2019-08-15" +
+                                         "&resource=https://storage.azure.com/",
+                                         headers: { "Metadata" => "true" })
+          .get
+          .body
+        return JSON.parse(access_key_request)["access_token"]
+      end
+
       def start
         super
-  
-        @bs = Azure::Storage::Blob::BlobService.create(storage_account_name: @azure_storage_account, storage_access_key: @azure_storage_access_key)
-  
-        ensure_container
+        if @azure_use_managed_identity
+          token_credential = Azure::Storage::Common::Core::TokenCredential.new get_access_token
+          token_signer = Azure::Storage::Common::Core::Auth::TokenSigner.new token_credential
+          @bs = Azure::Storage::Blob::BlobService.new(storage_account_name: @azure_storage_account, signer: token_signer)
 
-        @azure_storage_path = ''
-        @last_azure_storage_path = ''
+          # Refresh internal is 50 minutes
+          refresh_interval = 50 * 60
+          # The user-defined thread that renews the access token
+          cancelled = false
+          renew_token = Thread.new do
+            Thread.stop
+            while !cancelled
+              sleep(refresh_interval)
+              # Update the access token to the credential
+              token_credential.renew_token get_access_token
+            end
+          end
+          sleep 0.1 while renew_token.status != "sleep"
```

Comment:
Can you please also explain what this is doing?

Reply:
See previous answer, I can't really speak to this

Reply:
`Thread.stop` puts the thread to sleep so resumes the execution after